### PR TITLE
Feature/doc blocks

### LINKED ISSUE #13: [gadget] phpdoc

### PR DESCRIPTION
**Pull request by larsborn**
fix #13

### CODE REVIEW

File: src/Gadget/DocBlockGadget.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * @param ArrayNodeDefinition $node
+     */
+    public function configure(ArrayNodeDefinition $node)
+    {
+        $node->children()
+             ->node('files', 'paths')
+             ->defaultValue(['./'])
+             ->end()
+             ->node('missing_docblock', 'scalar')
```
Comment (DavidBadura) [inline src/Gadget/DocBlockGadget.php]: you can use `level` as node type :wink:

Comment (DavidBadura) [inline src/Gadget/DocBlockGadget.php]: see https://github.com/simpspector/analyser/blob/master/src/Config/LevelNodeDefinition.php

**Comment by DavidBadura** [comment]
don't forget again to register the gadget :wink: 
https://github.com/simpspector/analyser/blob/master/src/AnalyserBuilder.php#L178

**Comment by DavidBadura** [comment]
btw https://github.com/Block8/php-docblock-checker/blob/master/PhpDocblockChecker/CheckerCommand.php

File: src/Gadget/DocBlockGadget.php (hunk 1 above)
Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: `leve` does not allow `null` as value. We can allow this one, or create `NullableLevelNodeDefintion` or something else?

I think it would be the most conscious choice to set the level `missing_docblock: null` if one wants to disable this feature.

**Comment by larsborn** [comment]
me forget to register a Gadget?! Never!

**Comment by larsborn** [comment]
the PhpDocblockChecker looks very promising. It uses PHP_Token_Stream (and so will I) from phpunit. So well' add a part of phpunit as a non-dev-dependency now ;-)

File: src/Gadget/DocBlockGadget.php (hunk 1 above)
Comment (DavidBadura) [inline src/Gadget/DocBlockGadget.php]: i think you can use `canBeDisabled()`? http://symfony.com/doc/current/components/config/definition.html#optional-sections

in other case, yes we should define a new node definition for it.

Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: If I understand it right, `canBeDisabled` is only for config sections and not for single nodes

**Comment by larsborn** [comment]
I am not sure, how many feature I should implement. There is stuff like 
* an option to force an newline before a @return in a docstring
* detect return type (might be hard)
* also test functions which are in global namespace

But I think, I'll first leave it like this and only refactor the monster class `DocBlockGadget`. Any suggestions so far?

**Comment by tolry** [comment]
:+1: for starting with a "non-complete" set of features and also merge as soon as its usable. Whatever features are missing from the gadget can be put in separate issues, imho.

I will take a look at the code as soon as I get the time.

File: src/AnalyserBuilder.php [hunk 2]
```diff
@@ -185,5 +185,6 @@ protected function registerDefaultGadgets(RepositoryInterface $repository)
         $repository->add(new Gadget\CommentBlacklistGadget());
         $repository->add(new Gadget\FunctionBlacklistGadget());
         $repository->add(new Gadget\SecurityCheckerGadget($binDir . 'security-checker'));
+        $repository->add(new Gadget\DocBlockGadget());
```
Comment (tolry) [inline src/AnalyserBuilder.php]: not specific to this PR, but how about making the repository::add() function fluid?

File: src/Config/NullableLevelNodeDefinition.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+     */
+    public function __construct($name, NodeParentInterface $parent = null)
+    {
+        parent::__construct($name, $parent);
+
+        $this->values([
+            null,
+            Issue::LEVEL_CRITICAL,
+            Issue::LEVEL_ERROR,
+            Issue::LEVEL_NOTICE,
```
Comment (tolry) [inline src/Config/NullableLevelNodeDefinition.php]: switch notice and warning? :-)

File: src/Gadget/DocBlockGadget.php [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+class DocBlockGadget implements GadgetInterface
+{
+    /**
+     * @param ArrayNodeDefinition $node
+     */
+    public function configure(ArrayNodeDefinition $node)
+    {
+        $node->children()
+            ->node('files', 'paths')->defaultValue(['./'])->end()
+            ->node('missing_docblock', 'nullable_level')->defaultValue(Issue::LEVEL_ERROR)->end()
```
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: I like the nullable_level type, great addition - should we check the other gadget's configurations for other options that could make use of this? (seperate issue)

File: src/Gadget/DocBlockGadget.php [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+     *
+     * @return Result
+     */
+    public function run($path, array $options, AbstractLogger $logger)
+    {
+        $result = new Result();
+        $files  = FilesystemHelper::findFiles($path, $options['files']);
+        foreach ($files as $file) {
+            foreach ((new \PHP_Token_Stream($file))->getClasses() as $className => $class) {
+                foreach ($class['methods'] as $method) {
```
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: Are there no checks we want to perform on the class' docblock (e.g. check for `@author`)?

File: src/Gadget/DocBlockGadget.php [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+     *
+     * @return Result
+     */
+    private function processMethod(array $options, array $method, $className, $file)
+    {
+        $result = new Result();
+        if ($method['docblock'] === null) {
+            if ($options['missing_docblock'] !== null) {
+                $result->addIssue($this->createIssueForMissingDocblock($options, $method, $className, $file));
+            }
```
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: by adding a `return $result;` after this line, you can ditch the else after and save one indentation level

File: src/Gadget/DocBlockGadget.php [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+     * @param string $className
+     * @param string $file
+     *
+     * @return Result
+     */
+    private function processMethod(array $options, array $method, $className, $file)
+    {
+        $result = new Result();
+        if ($method['docblock'] === null) {
+            if ($options['missing_docblock'] !== null) {
```
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: do we want to add an issue here in any case, even if the function has no arguments and no return != null is done (hard to detect)? But this is probably something a later version might implement, right?

File: src/Gadget/DocBlockGadget.php [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        return $result;
+    }
+
+    /**
+     * @param array $tagsInDocblock
+     * @param FunctionSignature $signature
+     *
+     * @return array[]
+     */
+    private function inBoth(array $tagsInDocblock, FunctionSignature $signature)
```
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: Not sure about the naming here and the following two: this method creates an array of all method parameters, that also have a corresponding docblock annotation, right? The other ones return a list of parameters with missing docblocks respectively docblocks which have no corresponding method parameter.

Especially `inBoth()` is extremely generic and technical (you are referring to the two arrays that are given to the function).

first ideas

* filterParametersWithDocblocks
* filterParametersMissingDocblocks
* filterDocblocksMissingParameter

File: src/Gadget/DocBlockGadget.php [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    /**
+     * @param array $options
+     * @param array $method
+     * @param string $className
+     * @param string $file
+     *
+     * @return Issue
+     */
+    private function createIssueForMissingDocblock(array $options, array $method, $className, $file)
+    {
```
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: forgot to check if `$method['missing_docblock'] == null` and exit early? same goes for the following methods

File: tests/Gadget/DocBlockGadgetTest.php [hunk 10]
```diff
@@ -0,0 +1,161 @@
+<?php
+
+namespace Gadget;
+
+use SimpSpector\Analyser\Gadget\DocBlockGadget;
+use SimpSpector\Analyser\Issue;
+use SimpSpector\Analyser\Logger\NullLogger;
+use SimpSpector\Analyser\Tests\Gadget\GadgetTestCase;
+
+class DocBlockGadgetTest extends GadgetTestCase
```
Comment (tolry) [inline tests/Gadget/DocBlockGadgetTest.php]: maybe add tests to check for null options being honored?

File: src/Gadget/DocBlockGadget.php (hunk 9 above)
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: my mistake, you're checking this before even generating the issues

**Comment by tolry** [comment]
@larsborn that's all I could find for now - looks great! I really like being able to disable certain test, e.g. missing_docblocks :smiley:

**Comment by tolry** [comment]
btw: I think this feature set is more than enough for a first release

File: src/Config/NullableLevelNodeDefinition.php (hunk 3 above)
Comment (DavidBadura) [inline src/Config/NullableLevelNodeDefinition.php]: it is alphabetically ;-)

File: src/AnalyserBuilder.php (hunk 2 above)
Comment (DavidBadura) [inline src/AnalyserBuilder.php]: `Repository` implements a `RepositoryInterface` and i dont know it is a good idea to force the fluid pattern.

File: src/Config/NullableLevelNodeDefinition.php (hunk 3 above)
Comment (tolry) [inline src/Config/NullableLevelNodeDefinition.php]: if you switch notice with warning, you get ORDER by severity DESC, which makes sense to me - I find this sort order highly confusing, since it looks at first as if you are sorting by severity :smile:

File: src/AnalyserBuilder.php (hunk 2 above)
Comment (tolry) [inline src/AnalyserBuilder.php]: I think it would fit in this case and would not complicate this for anyone who wants to implement the interface, right? The only problem being of course that you can not force return types in php (yet), so this would linger in the interface's docblock, which is not a very strict contract. I'm fine either way, just thought this would make the gadget registration a bit more elegant.

**Comment by DavidBadura** [comment]
it is a bug right?

~~~
### obsolete variable AbstractLogger $logger in docblock of "Analyser::analyse($path, $configFile = null, AbstractLogger $logger = null)" on line 41

```php file:src/Analyser.php line:41 offset:36
     * @param string $path
     * @param string|null $configFile
     * @param AbstractLogger $logger
     * @return Result
     */
    public function analyse($path, $configFile = null, AbstractLogger $logger = null)
    {
        $logger = $logger ?: new NullLogger();
        $path   = rtrim($path, '/');
        if (!$configFile) {

~~~

https://travis-ci.org/simpspector/analyser/jobs/66134936#L276

**Comment by DavidBadura** [comment]
I don't know what to tell me.

~~~
### type of $configFile in docblock of "Analyser::analyse($path, $configFile = null, AbstractLogger $logger = null)" is string|null on line 41

```php file:src/Analyser.php line:41 offset:36
     * @param string $path
     * @param string|null $configFile
     * @param AbstractLogger $logger
     * @return Result
     */
    public function analyse($path, $configFile = null, AbstractLogger $logger = null)
    {
        $logger = $logger ?: new NullLogger();
        $path   = rtrim($path, '/');
        if (!$configFile) {
```

https://travis-ci.org/simpspector/analyser/jobs/66134936#L324

**Comment by DavidBadura** [comment]
you should write tests for it :D
btw. the right message should be `missing variable $loader in docblock [...]`

~~~
### missing variable LoaderInterface in docblock of "Analyser::__construct(ExecutorInterface $executor, LoaderInterface $loader)" on line 29

```php file:src/Analyser.php line:29 offset:24
    /**
     * @param ExecutorInterface $executor
     * @param LoaderInterface $loader
     */
    public function __construct(ExecutorInterface $executor, LoaderInterface $loader)
    {
        $this->executor = $executor;
        $this->loader   = $loader;
    }
```

https://travis-ci.org/simpspector/analyser/jobs/66134936#L356

**Comment by DavidBadura** [comment]
and good conclusion, your gadget has some issues :wink: 

https://travis-ci.org/simpspector/analyser/jobs/66134936#L2906

File: src/Config/NullableLevelNodeDefinition.php (hunk 3 above)
Comment (larsborn) [inline src/Config/NullableLevelNodeDefinition.php]: I ordered by severity now.

File: src/Gadget/DocBlockGadget.php (hunk 4 above)
Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: created #40

File: src/Gadget/DocBlockGadget.php (hunk 5 above)
Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: no there are none yet. I'll gather feature ideas in new issues

File: src/Gadget/DocBlockGadget.php (hunk 6 above)
Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: done.

File: src/Gadget/DocBlockGadget.php (hunk 7 above)
Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: yes, detecting return type might be very hard (I think one needs at least a full syntax tree for this). Die Option `missing_docblock` should enforce a docblock on *every* method or function. Even if its empty. Or didn't I get your questions right?

File: src/Gadget/DocBlockGadget.php (hunk 8 above)
Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: yes, this is one of the things, that I want to refactor before the merge.

File: tests/Gadget/DocBlockGadgetTest.php (hunk 10 above)
Comment (larsborn) [inline tests/Gadget/DocBlockGadgetTest.php]: I wrote one test for your favorite null option. I'll create an issue for more tests.

**Comment by larsborn** [comment]
concerning the bugs: I'll write tests and fix them.

File: src/Gadget/DocBlockGadget.php (hunk 7 above)
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: you got my question right, let me give you an example

```php
public function veryGoodFunctionNameLeavingNothingToDescribeInAComment()
{
 // ... some code without any return statements
}
```

In this case a docblock comment might be useless, no param or return entries in the docblock make sense. This gadget might result in the developer creating an empty docblock. But as you said we can only identify these cases if we check for return statements, therefore postponed?

Comment (larsborn) [inline src/Gadget/DocBlockGadget.php]: I think I understand the reason, why it is so important for you to disable the missing docblock feature. Only having

/**
  *
  */

seems kind of weird, but I like it to have a visual queue for "here begins a new function". So I would say, this is a matter of taste (and configuration).

#43 addresses the return statement topic.

**Comment by DavidBadura** [comment]
i think we should move this algorithm in there own repository and the gadget use this :-)

**Comment by DavidBadura** [comment]
you should add your gadget in `config/services.yml` :wink:

File: src/Gadget/DocBlockGadget.php (hunk 7 above)
Comment (tolry) [inline src/Gadget/DocBlockGadget.php]: Agreed, this is definitely a matter of preference. As I said before, this absolutely no show stopper, I just wanted to address the issue. Thanks for creating all the additional issues btw :+1:

**Comment by DavidBadura** [comment]
status?

**Comment by larsborn** [comment]
Current status:

* rebased against current master
* created #54
* created test for the obvious bugs

my next step is, obviously, fixing the bugs. Is there a reason you are asking? :-)

File: composer.json [hunk 11]
```diff
@@ -27,7 +27,11 @@
         "simpspector/twig-lint": "~1.1",
         "davidbadura/markdown-builder": "~1.0.3",
         "jms/serializer": "0.16.*",
-        "webmozart/path-util": "^2.0"
+        "webmozart/path-util": "^2.0",
+        "symfony/config": "~2.6",
+        "jms/serializer": "0.16.*",
```
Comment (DavidBadura) [inline composer.json]: dupplicated requirements

File: config/services.yml [hunk 12]
```diff
@@ -64,12 +64,17 @@ services:
             - { name: "simpspector.analyser.gadget" }
 
      simpspecotr.analyser.gadget.pdepend:
-        class: "SimpSpector\Analyser\Gadget\\PdependGadget"
+        class: "SimpSpector\Analyser\Gadget\PdependGadget"
```
Comment (DavidBadura) [inline config/services.yml]: revert this please, you broke it :wink:

**Comment by DavidBadura** [comment]
you should squash your commits :wink:

File: config/services.yml (hunk 12 above)
Comment (tolry) [inline config/services.yml]: Would this work with single quotes in yml without the escaping? The "arbitrary" escaping of only one segment of the path might confuse others too.

**Comment by larsborn** [comment]
I gave it a thought and I don't care if I squash my feature into a single commit before the merge or not as long as I may use as many commits as I want during development and don't have to divide it into "nice" pieces before the merge. So if you like it squashed, just write another +1 and I'll do it right before the merge.

**Comment by DavidBadura** [comment]
btw. reasons for squashing commits:

* A cleaner, simpler history.
* A smaller history. Less baggage when cloning repos.
* Removes "junk" commits, such as typo fixes.
* Gives the opportunity to improve commit messages.

-----------

seems still not to be working, see travis

**Comment by larsborn** [comment]
reasons against squashing:

* missing / artificial history
* obfuscation of the development process (it is not visible to others how the development exactly happened)
* additional work (especially in the case where you craft an artificial development history)

I know that I'm not done yet :-p.

**Comment by larsborn** [comment]
.. I think, this is ready to merge. This one process-method is a bit big, but I don't have any bright ideas at the moment ;-)

**Comment by larsborn** [comment]
@simpspector/owners

File: src/Gadget/DocBlockGadget.php (hunk 8 above)
Comment (DavidBadura) [inline src/Gadget/DocBlockGadget.php]: ping

File: src/Gadget/DocBlockGadget.php [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+        $className,
+        $file,
+        FunctionParameter $functionParameter,
+        $docBlockType
+    )
+    {
+        $issue = new Issue(
+            $this,
+            sprintf(
+                '"%s" variable %s has type %s in docblock',
```
Comment (DavidBadura) [inline src/Gadget/DocBlockGadget.php]: I find the error message saying nothing.

```
"%s" variable %s has a type missmatch %s in docblock, expected %s
```

**Comment by tolry** [comment]
:+1:  for merge

As for the complexity issues with the function processMethod() - at first glance it seems to me that it would be an option to extract the three if blocks into separate methods with an early exit if the respective option is not set

``` php
$this->processMethodDocblock(...);
$this->processMethodTypeMismatch(...);
$this->processMethodObsoleteVariable(...);
$this->processMethodMissingVariable(...);
```

Although you might porbably have to pass lots of parameters around - this I haven't checked yet.

**Comment by larsborn** [comment]
ready for review.

**Comment by DavidBadura** [comment]
+1

**Comment by dragonito** [comment]
-1 ;)
